lightning-c-bindings/README.md

```diff
@@ -0,0 +1,143 @@
+
+All of the Rust-Lightning types are mapped into C equivalents which take a few forms, namely:
+
+ * Structs are mapped into a simple wrapper containing a pointer to the native Rust-Lightning
+   object and a flag to indicate whether the object is owned or only a reference. Such mappings
+   usually generate a `X_free` function which must be called to release the allocated resources.
+   Note that calling `X_free` will do nothing if the underlying pointer is NULL or if the reference
+   flag is set.
+
+   You MUST NOT create such wrapper structs manually, relying instead on constructors which have
+   been mapped from equivalent Rust constructors.
+
+   Note that, thanks to the is-reference flag, such structs effectively represent both `&RustThing`
+   and `RustThing`. Further, the same applies to `Option<RustThing>`, in which cases `inner` may be
+   set to NULL.
+
+   For example, this is the mapping of ChannelManager.
+   ```
+   typedef struct MUST_USE_STRUCT LDKChannelManager {
+      /** ... */
+      LDKlnChannelManager *inner;
+      bool _underlying_ref;
+   } LDKChannelManager;
+   ```
+
+ * Traits are mapped into a concrete struct containing a void pointer and a jump table listing the
+   functions which the trait must implement. The void pointer may be set to any value and is never
+   interpreted (or dereferenced) by the bindings logic in any way. It is passed as the first
+   argument to all function calls in the trait. You may wish to use it as a pointer to your own
+   internal data structure, though it may also occasionally make sense to e.g. cast a file
+   descriptor into a void pointer and use it to track a socket.
+
+   Each trait additionally contains a `free` and `clone` function pointer, which may be NULL. The
+   `free` function is passed the void pointer when the object is `Drop`ed in Rust. The `clone`
+   function is passed the void pointer when the object is `Clone`ed in Rust, returning a new void
+   pointer for the new object.
+
+   For example, `LDKSocketDescriptor` is mapped as follows:
+   ```
+   typedef struct LDKSocketDescriptor {
+      void *this_arg;
+      /** ... */
+      uintptr_t (*send_data)(void *this_arg, LDKu8slice data, bool resume_read);
+      /** ... */
+      void (*disconnect_socket)(void *this_arg);
+      bool (*eq)(const void *this_arg, const void *other_arg);
+      uint64_t (*hash)(const void *this_arg);
+      void *(*clone)(const void *this_arg);
+      void (*free)(void *this_arg);
+   } LDKSocketDescriptor;
+   ```
+
+ * Rust structs that implement a trait result in the generation of an `X_as_Y` function which allows
+   you to use the native Rust object in place of the trait. Such generated objects are only valid as
+   long as the original Rust native object is owned by a C-wrapped struct, and has not been `free`'d
+   or moved as a part of a Rust function call.
+
+ * Rust "unitary" enums are mapped simply as an equivalent C enum, however some Rust enums have
```

Comment:
I think an example of a value enum and a Sentinel pattern output would be helpful here.